src/porepy/numerics/ad/operators.py

```diff
@@ -275,63 +325,91 @@ def parse(self, mdg: pp.MixedDimensionalGrid) -> Any:
+                (and possibly derivatives as identities) of the global vector at current
+                time and iterate.
+
+        Returns:
+            The numerical representation of this operator.
+
         """
 
         # The parsing strategy depends on the operator at hand:
-        # 1) If the operator is a Variable, it will be represented according to its
-        #    stored state.
-        # 2) If the operator is a leaf in the tree-representation of the operator,
-        #    parsing is left to the operator itself.
-        # 3) If the operator is formed by combining other operators lower in the tree,
-        #    parsing is handled by first evaluating the children (leads to recursion)
-        #    and then perform the operation on the result.
-
-        # Check for case 1 or 2
-        if isinstance(op, pp.ad.Variable) or isinstance(op, Variable):
-            # Case 1: Variable
-
-            # How to access the array of (Ad representation of) states depends on
-            # whether this is a single or combined variable; see self.__init__,
-            # definition of self._variable_ids.
-            # TODO: no difference between merged or no mixed-dimensional variables!?
-            if isinstance(op, pp.ad.MixedDimensionalVariable) or isinstance(
-                op, MixedDimensionalVariable
-            ):
-                if op.prev_time:
-                    return self._prev_vals[op.id]
-                elif op.prev_iter:
-                    return self._prev_iter_vals[op.id]
-                else:
-                    return self._ad[op.id]
-            else:
-                if op.prev_time:
-                    return self._prev_vals[op.id]
-                elif op.prev_iter or not (
-                    op.id in self._ad
-                ):  # TODO make it more explicit that op corresponds to a non_ad_variable?
-                    # e.g. by op.id in non_ad_variable_ids.
-                    return self._prev_iter_vals[op.id]
-                else:
-                    return self._ad[op.id]
-        elif isinstance(op, pp.ad.AdArray):
-            # When using nested operator functions, op can be an already evaluated term.
-            # Just return it.
+        # 1) If it is numeric, then it is already some sort of leaf. Return it
+        # 2) If it is a leaf
+        #    a) A md-variable with dofs per atomic variable
+        #    b) An atomic variable, with its dofs
+        #    c) Some wrapper for discretizations or other data.
+        # 3) If it is a non-variable operator with children, invoke recursion and
+        #    proceed to the non-void operation.
+
+        # Case 1), Some numeric data, or already evaluated operator.
+        if isinstance(op, AdArray | np.ndarray | pp.number):
             return op
 
-        elif op.is_leaf():
-            # Case 2
-            return op.parse(mdg)  # type:ignore
+        # to continue, we must assert it is an actual, unparsed operator
```

Comment:
Operator parsing changed. Usage of Variable.parse is introduced to handle arbitrary iter and time step depth.
Also, slicing of the global AdArray is introduced.
Overall more efficient and clearer code.